theory/DD/countpairs_kernels.c.src

```diff
@@ -19,6 +19,281 @@
+          const NEON_FLOATS m_y1 = NEON_LOAD_FLOATS(localy1);
+          const NEON_FLOATS m_z1 = NEON_LOAD_FLOATS(localz1);
+
+          localx1 += NEON_NVEC;
+          localy1 += NEON_NVEC;
+          localz1 += NEON_NVEC;
+
+          for(int w = 0; w < pair.num_weights; w++){
+              pair.weights1[w].n = NEON_LOAD_FLOATS(local_w1.weights[w]);
+              local_w1.weights[w] += NEON_NVEC;
+          }
+
+          const NEON_FLOATS m_max_dz = NEON_SET_FLOAT(max_dz);
+          const NEON_FLOATS m_sqr_rpmax = NEON_SET_FLOAT(sqr_rpmax);
+          const NEON_FLOATS m_sqr_rpmin = NEON_SET_FLOAT(sqr_rpmin);
+
+          const NEON_FLOATS m_xdiff = NEON_SUBTRACT_FLOATS(m_x1, m_xpos);  //(x[j] - x0)
+          const NEON_FLOATS m_ydiff = NEON_SUBTRACT_FLOATS(m_y1, m_ypos);  //(y[j] - y0)
+          const NEON_FLOATS m_zdiff = NEON_SUBTRACT_FLOATS(m_z1, m_zpos);  //z2[j:j+NVEC-1] - z1
+
+          const NEON_UINTS m_mask_geq_pimax = NEON_COMPARE_FLOATS_GE(m_zdiff,m_max_dz);
+          if(NEON_TEST_COMPARISON(m_mask_geq_pimax) > 0) {
+              j = N1;//but do not break yet, there might be pairs in this chunk
+          }
+
+        //   if(need_weightavg){
+              pair.dx.n = m_xdiff;
+              pair.dy.n = m_ydiff;
+              pair.dz.n = m_zdiff;
+        //   }
+#ifdef __ARM_FEATURE_FMA
+          NEON_FLOATS r2 = NEON_SQUARE_FLOAT(m_xdiff);
+          r2 = NEON_FMA_FLOATS(r2, m_ydiff, m_ydiff);
+          r2 = NEON_FMA_FLOATS(r2, m_zdiff, m_zdiff);
+#else
+          NEON_FLOATS r2 = NEON_ADD_FLOATS(NEON_ADD_FLOATS(NEON_SQUARE_FLOAT(m_xdiff), NEON_SQUARE_FLOAT(m_ydiff)), 
+                                           NEON_SQUARE_FLOAT(m_zdiff));
+#endif
+
+          NEON_UINTS m_mask_left;
+          const NEON_UINTS m_rpmin_mask = NEON_COMPARE_FLOATS_GE(r2, m_sqr_rpmin);
+          const NEON_UINTS m_rpmax_mask = NEON_COMPARE_FLOATS_LT(r2, m_sqr_rpmax);
+          m_mask_left = NEON_BITWISE_AND(m_rpmin_mask, m_rpmax_mask);
+          if(NEON_TEST_COMPARISON(m_mask_left) == 0) {
+              continue;
+          }
+          r2 = NEON_BLEND_FLOATS_WITH_MASK(m_sqr_rpmax, r2, m_mask_left);
+
+        //   if(need_rpavg) {
+              union_mDperp.m_Dperp = NEON_SQRT_FLOAT(r2);
+        //   }
+          if(need_weightavg){
+              union_mweight.m_weights = neon_weight_func(&pair);
+          }
+
+          NEON_INTS m_rpbin = NEON_SET_INT(0);
+          for(int kbin=nbin-1;kbin>=1;kbin--) {
+              const NEON_UINTS m1 = NEON_COMPARE_FLOATS_GE(r2,m_rupp_sqr[kbin-1]);
+              const NEON_UINTS m_bin_mask = NEON_BITWISE_AND(m1,m_mask_left);
+              m_mask_left = NEON_COMPARE_FLOATS_LT(r2,m_rupp_sqr[kbin-1]);
+              const int test2  = NEON_TEST_COMPARISON(m_bin_mask);
+              npairs[kbin] += NEON_BIT_COUNT_INT(test2);
+              if(need_rpavg || need_weightavg){
+                  m_rpbin = NEON_BLEND_INTS_WITH_MASK(m_rpbin, NEON_SET_INT(kbin), m_bin_mask);
```

Comment:
Everything looks good to me; I'm not seeing why `rpavg` and `weightavg` would be wrong. This line stands out as one place things could go wrong, but reading the documentation of these instructions, it looks okay to me.